Mathlib/Dynamics/TopologicalEntropy/DynamicalUniformity.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2024 Damien Thomine. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damien Thomine, Pietro Monticone
+-/
+import Mathlib.Order.Interval.Finset.Nat
+import Mathlib.Order.OmegaCompletePartialOrder
+import Mathlib.Topology.UniformSpace.Basic
+
+/-!
+# Dynamical uniformities
+Bowen-Dinaburg's definition of topological entropy of a transformation `T` in a metric space
+`(X, d)` relies on the so-called dynamical balls. These balls are sets
+`B (x, ε, n) = { y | ∀ k < n, d(T^[k] x, T^[k] y) < ε }`.
+
+We implement Bowen-Dinaburg's definitions in the more general context of uniform spaces. Dynamical
+balls are replaced by what we call dynamical uniformities. This file collects all general lemmas
+about these objects.
+
+## Main definitions
+- `DynamicalUni`: dynamical uniformity associated with a given transformation `T`, uniformity `U`
+and time `n`.
+
+## Tags
+entropy
+
+## TODO
+Once #PR14718 has passed, add product of uniformities.
+
+In the context of (pseudo-e)metric spaces, relate the usual definition of dynamical balls with
+these dynamical uniformities.
+-/
+
+namespace DynamicalUniformity
+
+open Prod Set Uniformity UniformSpace
+
+variable {X : Type*}
+
+/-- A dynamical uniform neighborhood is the uniform space version of dynamical balls.-/
+def DynamicalUni (T : X → X) (U : Set (X × X)) (n : ℕ) : Set (X × X) :=
+  ⋂ k < n, (map T T)^[k] ⁻¹' U
+
+theorem dynamical_uni_inter_Ico (T : X → X) (U : Set (X × X)) (n : ℕ) :
+    DynamicalUni T U n = ⋂ k : Ico 0 n, (map T T)^[k] ⁻¹' U := by
+  simp [DynamicalUni, iInter_coe_set, mem_Ico, zero_le, true_and]
+
+theorem dynamical_uni_mem {T : X → X} {U : Set (X × X)} {n : ℕ} {x y : X} :
+    (x, y) ∈ DynamicalUni T U n ↔ ∀ k < n, (T^[k] x, T^[k] y) ∈ U := by
+  simp [DynamicalUni, map_iterate, mem_preimage, mem_iInter, map_apply]
+
+theorem dynamical_balls_mem {T : X → X} {U : Set (X × X)} {n : ℕ} {x y : X} :
+    y ∈ ball x (DynamicalUni T U n) ↔ ∀ k < n, T^[k] y ∈ ball (T^[k] x) U := by
+  simp [ball, mem_preimage]; exact dynamical_uni_mem
+
+theorem dynamical_uni_of_uni [UniformSpace X] {T : X → X} (h : UniformContinuous T)
+    {U : Set (X × X)} (U_uni : U ∈ 𝓤 X) (n : ℕ) :
+    DynamicalUni T U n ∈ 𝓤 X := by
+  rw [dynamical_uni_inter_Ico T U n]
+  refine Filter.iInter_mem.2 fun k ↦ ?_
+  rw [map_iterate T T k]
+  exact uniformContinuous_def.1 (UniformContinuous.iterate T k h) U U_uni
+
+theorem dynamical_uni_of_rfl (T : X → X) {U : Set (X × X)} (h : idRel ⊆ U) (n : ℕ) :
```

Comment:
```suggestion
theorem idRel_subset_dynEntourage (T : X → X) {U : Set (X × X)} (h : idRel ⊆ U) (n : ℕ) :
```
Don't try to be creative with names, and follow the sequence of symbols in the name unless there is a good reason not to.